confmap/expand.go

```diff
@@ -0,0 +1,206 @@
+			return "", false, errURILimit
+		}
+
+		uri := findURI(v)
+		if uri != "" && uri == value {
+			// If the value is a single URI, then the return value can be anything.
+			// This is the case `foo: ${file:some_extra_config.yml}`.
+			return mr.expandURI(ctx, v)
+		}
+
+		// Embedded or nested URIs.
+		return mr.findAndExpandURI(ctx, v)
+	case []any:
+		nslice := make([]any, 0, len(v))
+		nchanged := false
+		for _, vint := range v {
+			val, changed, err := mr.expandValue(ctx, vint)
+			if err != nil {
+				return nil, false, err
+			}
+			nslice = append(nslice, val)
+			nchanged = nchanged || changed
+		}
+		return nslice, nchanged, nil
+	case map[string]any:
+		nmap := map[string]any{}
+		nchanged := false
+		for mk, mv := range v {
+			val, changed, err := mr.expandValue(ctx, mv)
+			if err != nil {
+				return nil, false, err
+			}
+			nmap[mk] = val
+			nchanged = nchanged || changed
+		}
+		return nmap, nchanged, nil
+	}
+	return value, false, nil
+}
+
+// findURI finds the URI corresponding to the first closing bracket in input. It returns
+// the URI if it is expandable, or an empty string if it is not expandable.
+// Note: findURI is only called when input contains a closing bracket.
+func findURI(input string) string {
+	closeIndex := strings.Index(input, "}")
+	openIndex := strings.LastIndex(input[:closeIndex+1], "${")
+	if openIndex < 0 {
+		// Should not expand because there is a missing ${.
+		return ""
+	}
+
+	uri := input[openIndex : closeIndex+1]
+
+	if !strings.Contains(uri, ":") {
+		// Should not expand. This is expanded in the expandconverter.
+		return ""
+	}
+
+	return uri
+}
+
+// findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input.
+func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (output string, changed bool, err error) {
+	var repl string
+	var expandedRemaining string
+
+	uri := findURI(input)
+	if uri == "" {
+		// The first URI in input is not expandable. Strip the first URI from input and check if
+		// other URIs are expandable.
+		closeIndex := strings.Index(input, "}")
+		noExpand := input[:closeIndex+1]
+		remaining := input[closeIndex+1:]
+
+		// if remaining does not contain }, there are no URIs left: stop recursion.
+		if !strings.Contains(remaining, "}") {
+			return input, changed, err
+		}
+
+		expandedRemaining, changed, err = mr.findAndExpandURI(ctx, remaining)
+		return noExpand + expandedRemaining, changed, err
+	}
+	repl, changed, err = mr.expandStringURI(ctx, uri)
+	input = strings.ReplaceAll(input, uri, repl)
+	return input, changed, err
+}
+
+// expandStringURI tries to expand uri as a string. If an expandable URI is found and it can be converted to a string,
+// it returns the expanded uris string value, true and nil. Otherwise, it returns the unchanged input, false and the expanding error.
+func (mr *Resolver) expandStringURI(ctx context.Context, uri string) (string, bool, error) {
```

Comment:
If I see correctly, only one caller of this. Consider to make this just a "toString" func and move the call to `expandURI` to the caller func.

Reply:
Changed to `toString` func.